### DISCUSSION TITLE
Symfony 6 API http request circular reference problem

### DISCUSSION THREAD

**Discussion by llaumegui27** [symfony/symfony#50642] [Q&A]
I have a problem when I try to use my APPI to populate the reservations table in my database. This table contains 2 foreign keys (relationships that refer to the id of my user table and the id of my room table in ManyToOne for both).  

In my addReservation() function in my ReservationController.php page i have this error when i try to POST my request : 
"A circular reference has been detected when serializing the object of class &quot;App\Entity\User&quot; (configured limit: 1). (500 Internal Server Error)".

My Rervation entity : 
`<?php

namespace App\Entity;

use App\Repository\ReservationRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: ReservationRepository::class)]
class Reservation
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(type: Types::DATETIME_MUTABLE)]
    private ?\DateTimeInterface $dateHeureDebut = null;

    #[ORM\Column(type: Types::DATETIME_MUTABLE)]
    private ?\DateTimeInterface $dateHeureFin = null;

    #[ORM\Column]
    private ?bool $etat = null;

    #[ORM\ManyToOne(inversedBy: 'userReservation')]
    private ?User $idUser = null;

    #[ORM\ManyToOne(inversedBy: 'roomReservation')]
    private ?Room $idRoom = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getDateHeureDebut(): ?\DateTimeInterface
    {
        return $this->dateHeureDebut;
    }

    public function setDateHeureDebut(\DateTimeInterface $dateHeureDebut): static
    {
        $this->dateHeureDebut = $dateHeureDebut;

        return $this;
    }

    public function getDateHeureFin(): ?\DateTimeInterface
    {
        return $this->dateHeureFin;
    }

    public function setDateHeureFin(\DateTimeInterface $dateHeureFin): static
    {
        $this->dateHeureFin = $dateHeureFin;

        return $this;
    }

    public function isEtat(): ?bool
    {
        return $this->etat;
    }

    public function setEtat(bool $etat): static
    {
        $this->etat = $etat;

        return $this;
    }

    public function getIdUser(): ?User
    {
        return $this->idUser;
    }

    public function setIdUser(?User $idUser): static
    {
        $this->idUser = $idUser;

        return $this;
    }

    public function getIdRoom(): ?Room
    {
        return $this->idRoom;
    }

    public function setIdRoom(?Room $idRoom): static
    {
        $this->idRoom = $idRoom;

        return $this;
    }
}
`
My RerservationController : 
`
namespace App\Controller;

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Routing\Annotation\Route;
use App\Entity\Reservation;
use App\Repository\ReservationRepository;
use App\Entity\User;
use App\Repository\UserRepository;
use App\Entity\Room;
use App\Repository\RoomRepository;
use Doctrine\Persistence\ManagerRegistry;
use DateTimeImmutable;

class ReservationController extends AbstractController
{
    #[Route('/reservations', name: 'reservations')]
    public function reservations(ReservationRepository $reservationRepository): Response
    {
        $reservations = $reservationRepository->findAll();
        $responseData = [];

        foreach ($reservations as $reservation) {
            $responseData[] = [
                'id' => $reservation->getId(),
                'date_heure_debut' => $reservation->getDateHeureDebut(),
                'date_heure_fin' => $reservation->getDateHeureFin(),
                'id_user_id' => $reservation->getIdUser()->getId(),
                'id_room_id' => $reservation->getIdRoom()->getId()
            ];
        }
        return $this->json($responseData);
    }

    #[Route('/add-reservation', name: 'add-reservation', methods: ['POST'])]
    public function addReservation(Request $request, ManagerRegistry $doctrine): Response
    {

        $entityManager = $doctrine->getManager();
        $data = json_decode($request->getContent(), true);
        $reservation = new Reservation();

        $idUser = $entityManager->getRepository(User::class)->find($data['id_user_id']);
        $idRoom = $entityManager->getRepository(Room::class)->find($data['id_room_id']);

        $dateHeureDebut = new DateTimeImmutable($data['date_heure_debut']);
        $dateHeureFin = new DateTimeImmutable($data['date_heure_fin']);

        $reservation->setIdUser($idUser);
        $reservation->setIdRoom($idRoom);
        $reservation->setDateHeureDebut($dateHeureDebut);
        $reservation->setDateHeureFin($dateHeureFin);
        $reservation->setEtat($data['etat']);

        $entityManager->persist($idUser);
        $entityManager->persist($idRoom);
        $entityManager->persist($reservation);
        $entityManager->flush();
        //return new JsonResponse('success');
        return $this->json($reservation, Response::HTTP_OK);
    }
}
`
My database : 
![image](https://github.com/symfony/symfony/assets/91952098/e3366224-3124-41c6-a539-b170d5e38448)

If someone can please help me,,
Thx.

**Comment by ghost** [comment]
Reservation has User field, User field probably has collection of Reservation (probably because you didn't provide code block for User class) that again has User field and so on and so on. There are many solutions to your problem depending on what response you expect. 
- you can exclude `Reservation::$idUser` or `User::$userReservation` from serialization
- you can return `Reservation::$idUser` as just id, not an object
- write custom normalizer for Reservation entity
- return `User::$userReservations` as an array of ids

Read this section about circural references https://symfony.com/doc/current/components/serializer.html#handling-circular-references

**Reply by llaumegui27** [reply]
Hi, thx for the reply, I tried a few things but nothing work, i will show you my User entity : 
`<?php

namespace App\Entity;

use App\Repository\UserRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: UserRepository::class)]
class User
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 255)]
    private ?string $name = null;

    #[ORM\Column(length: 255)]
    private ?string $mail = null;

    #[ORM\Column(length: 255)]
    private ?string $password = null;

    #[ORM\Column(nullable: true)]
    private ?bool $teacher = null;

    #[ORM\Column(nullable: true)]
    private ?bool $admin = null;

    #[ORM\Column(nullable: true)]
    private ?bool $superAdmin = null;

    #[ORM\OneToMany(mappedBy: 'idUser', targetEntity: Reservation::class)]
    private Collection $userReservation;

    public function __construct()
    {
        $this->userReservation = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(string $name): static
    {
        $this->name = $name;

        return $this;
    }

    public function getMail(): ?string
    {
        return $this->mail;
    }

    public function setMail(string $mail): static
    {
        $this->mail = $mail;

        return $this;
    }

    public function getPassword(): ?string
    {
        return $this->password;
    }

    public function setPassword(string $password): static
    {
        $this->password = $password;

        return $this;
    }

    public function isTeacher(): ?bool
    {
        return $this->teacher;
    }

    public function setTeacher(?bool $teacher): static
    {
        $this->teacher = $teacher;

        return $this;
    }

    public function isAdmin(): ?bool
    {
        return $this->admin;
    }

    public function setAdmin(?bool $admin): static
    {
        $this->admin = $admin;

        return $this;
    }

    public function isSuperAdmin(): ?bool
    {
        return $this->superAdmin;
    }

    public function setSuperAdmin(?bool $superAdmin): static
    {
        $this->superAdmin = $superAdmin;

        return $this;
    }

    /**
     * @return Collection<int, Reservation>
     */
    public function getUserReservation(): Collection
    {
        return $this->userReservation;
    }

    public function addUserReservation(Reservation $userReservation): static
    {
        if (!$this->userReservation->contains($userReservation)) {
            $this->userReservation->add($userReservation);
            $userReservation->setIdUser($this);
        }

        return $this;
    }

    public function removeUserReservation(Reservation $userReservation): static
    {
        if ($this->userReservation->removeElement($userReservation)) {
            // set the owning side to null (unless already changed)
            if ($userReservation->getIdUser() === $this) {
                $userReservation->setIdUser(null);
            }
        }

        return $this;
    }
}
`
And my Room entity : 
`<?php

namespace App\Entity;

use App\Repository\RoomRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: RoomRepository::class)]
class Room
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 255)]
    private ?string $name = null;

    #[ORM\Column(length: 255)]
    private ?string $subject = null;

    #[ORM\Column]
    private ?int $places = null;

    #[ORM\Column]
    private ?int $participants = null;

    #[ORM\OneToMany(mappedBy: 'idRoom', targetEntity: Reservation::class)]
    private Collection $roomReservation;

    public function __construct()
    {
        $this->roomReservation = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(string $name): static
    {
        $this->name = $name;

        return $this;
    }

    public function getSubject(): ?string
    {
        return $this->subject;
    }

    public function setSubject(string $subject): static
    {
        $this->subject = $subject;

        return $this;
    }

    public function getPlaces(): ?int
    {
        return $this->places;
    }

    public function setPlaces(int $places): static
    {
        $this->places = $places;

        return $this;
    }

    public function getParticipants(): ?int
    {
        return $this->participants;
    }

    public function setParticipants(int $participants): static
    {
        $this->participants = $participants;

        return $this;
    }

    /**
     * @return Collection<int, Reservation>
     */
    public function getRoomReservation(): Collection
    {
        return $this->roomReservation;
    }

    public function addRoomReservation(Reservation $roomReservation): static
    {
        if (!$this->roomReservation->contains($roomReservation)) {
            $this->roomReservation->add($roomReservation);
            $roomReservation->setIdRoom($this);
        }

        return $this;
    }

    public function removeRoomReservation(Reservation $roomReservation): static
    {
        if ($this->roomReservation->removeElement($roomReservation)) {
            // set the owning side to null (unless already changed)
            if ($roomReservation->getIdRoom() === $this) {
                $roomReservation->setIdRoom(null);
            }
        }

        return $this;
    }
}
`
I don't really understand this circular reference phenomen. Can you please tell me how to proceed?
Thx

**Reply by ghost** [reply]
I've explained this in the previous comment. Also added a link to the documentation about circular reference. There's nothing more I can say right now. 

For now you can add `#[Ignore]` to the `User:$userReservation` field to make the app working and remove it when you will understand circular reference problem.

**Reply by llaumegui27** [reply]
How can I return Reservation::$idUser as just id, not an object as you said ?

**Reply by ghost** [reply]
AFAIR by default SF serializer serializes public fields, private fields that have public accessors and also public accessor without fields. So you can `#[Ignore]` the `Reservation::$idUser` field and add custom getter like `getUserId()` that will return `$this->isUser->getId()`. 

Example:
```php
class A {
    private string $a = 'a';

    #[Ignore]
    private string $b = 'b';

    public function getA(): string
    {
        return $this->a;
    }

    public function getB(): string
    {
        return $this->b;
    }

    public function getC(): string
    {
        return 'c';
    }
}
```

serialized json will be:
```json
{"a":"a","c":"c"}
```

**Comment by llaumegui27** [comment]
No may bad, I'm not trying to return the idUser, just set the idUser. I can already return the idUser. 
Is it possible to create a custom set like the getUserId() you mentioned above? So that I can insert my userId in my database

**Reply by ghost** [reply]
I'm sorry, you comments are incomprehensible. Please proofread before you submit a comment.

I've helped you with the circular reference problem you mentioned in the discussion description. It you have another issue please open a new discussion.

**Comment by llaumegui27** [comment]
Ok my bad I'll try to explain it clearly I want to post (with postman) in a table of my MySQL database where there are two foreign keys idUser and idRoom which are ManyToOne relationships.
My json body : 
```
{
  "date_heure_debut": "2023-06-12T09:00:00",
  "date_heure_fin": "2023-06-12T12:30:00",
  "etat": 0,
  "id_user_id": 2,
  "id_room_id": 1
}
```

And i have this error : 
`A circular reference has been detected when serializing the object of class &quot;App\Entity\User&quot; (configured limit: 1). (500 Internal Server Error)`

And I just want to know how I'm going to get the insertion into my table based on the code sent earlier and I don't want to ignore these fields - I need them.

**Reply by ghost** [reply]
I've already said that the exception throws on the last line, $this->json() when sending back response (reservation is being serialized and you have to do what I said before to have this error gone). It's not related to what you described you want to happen. It should already work like you want because flush() is before the line that throws.
